File: packages/lynx/host/android/app/src/main/java/com/yee94/openchamber/lynx/OpenChamberLynxHttpClient.kt
package com.yee94.openchamber.lynx

/**
 * Host HTTP client — OkHttp shapes for Lynx inject.
 * Mirrors packages/lynx/src/host/httpClient.ts
 */
data class LynxHostHttpRequest(
    val url: String,
    val method: String = "GET",
    val headers: Map<String, String> = emptyMap(),
    val body: String? = null,
    val timeoutMs: Long = 8_000L,
)

data class LynxHostHttpResponse(
    val ok: Boolean,
    val status: Int,
    val headers: Map<String, String> = emptyMap(),
    val bodyText: String? = null,
)

interface LynxHttpClienting {
    suspend fun request(input: LynxHostHttpRequest): LynxHostHttpResponse?
}

/**
 * Stub OkHttp call factory. Replace [callFactory] when OkHttp / Lynx tunnel is linked.
 * Linux agents never claim a live network call from this stub alone.
 */
class OpenChamberLynxHttpClient(
    private val callFactory: (suspend (LynxHostHttpRequest) -> LynxHostHttpResponse?)? = null,
) : LynxHttpClienting {
    override suspend fun request(input: LynxHostHttpRequest): LynxHostHttpResponse? {
        // Inject point: OkHttpClient.newCall(Request.Builder()...).
        return callFactory?.invoke(input)
    }
}

File: packages/lynx/host/ios/OpenChamberLynxHttpClient.swift
import Foundation

/// Host HTTP client — URLSession shapes for Lynx inject.
/// Mirrors `packages/lynx/src/host/httpClient.ts`.
/// Wire URLSession (optional TLS pinning / cookie jar) before first Lynx paint.
struct OpenChamberLynxHttpRequest {
  var url: URL
  var method: String
  var headers: [String: String]
  var body: Data?
  var timeoutMs: Int
}

struct OpenChamberLynxHttpResponse {
  var ok: Bool
  var status: Int
  var headers: [String: String]
  var bodyText: String?
}

protocol OpenChamberLynxHttpClienting: AnyObject {
  func request(_ input: OpenChamberLynxHttpRequest) async -> OpenChamberLynxHttpResponse?
}

/// Stub: builds a real URLRequest shape; returns nil until URLSession is injected.
final class OpenChamberLynxHttpClient: OpenChamberLynxHttpClienting {
  var session: URLSession = .shared

  func request(_ input: OpenChamberLynxHttpRequest) async -> OpenChamberLynxHttpResponse? {
    var urlRequest = URLRequest(url: input.url, timeoutInterval: TimeInterval(max(1, input.timeoutMs)) / 1000)
    urlRequest.httpMethod = input.method
    for (key, value) in input.headers {
      urlRequest.setValue(value, forHTTPHeaderField: key)
    }
    urlRequest.httpBody = input.body
    // Inject point: replace with pinned session / relay tunnel.
    // Linux/Xcode-less agents never claim a live call from this stub alone.
    do {
      let (data, response) = try await session.data(for: urlRequest)
      guard let http = response as? HTTPURLResponse else { return nil }
      return OpenChamberLynxHttpResponse(
        ok: (200...299).contains(http.statusCode),
        status: http.statusCode,
        headers: http.allHeaderFields.reduce(into: [:]) { acc, pair in
          if let key = pair.key as? String, let value = pair.value as? String {
            acc[key] = value
          }
        },
        bodyText: String(data: data, encoding: .utf8)
      )
    } catch {
      return nil
    }
  }
}
